Implement safe alignment in flexbox.

fantasai opened the pull request:
#### 4258e7ee63a17d362a9d57f91615925b176962ca
<pre>
Implement safe alignment in flexbox.
<a href="https://bugs.webkit.org/show_bug.cgi?id=263893">https://bugs.webkit.org/show_bug.cgi?id=263893</a>
<a href="https://rdar.apple.com/118000717">rdar://118000717</a>

Reviewed by Sammy Gill.

Coverts alignment to Start when available space is negative and overflow alignment safety is on.

* LayoutTests/TestExpectations: Pass more tests!
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-002.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-003.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-004.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-005.html: Fix (and augment) incorrect test.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::initialJustifyContentOffset): Apply overflow safety to main-axis content alignment.
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild): Apply overflow safety to static positioning.
(WebCore::RenderFlexibleBox::overflowAlignmentForChild const): Implement new helper method.
(WebCore::initialAlignContentOffset): Apply overflow safety to main-axis content alignment.
(WebCore::RenderFlexibleBox::alignFlexLines): Don&apos;t skip alignment calculation when safety is on, pass safety value to initialAlignContentOffset().
(WebCore::RenderFlexibleBox::alignChildren): Apply overflow safety to cross-axis item alignment.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Apply overflow safety to overflow bounds.
* Source/WebCore/rendering/RenderFlexibleBox.h: Add helper method overflowAlignmentForChild(), parallel to alignmentForChild().

Canonical link: <a href="https://commits.webkit.org/274304@main">https://commits.webkit.org/274304@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/232226f41aee368505033eb4fbbd4b51a3c0f27d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38532 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17463 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/40860 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41070 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34213 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/40837 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20248 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/14808 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32398 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39105 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14711 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/12781 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42346 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/34995 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34786 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38599 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13376 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11057 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/36808 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/14976 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33681 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8668 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/13843 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14449 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1923,7 +1928,11 @@ LayoutUnit RenderFlexibleBox::staticMainAxisPositionForPositionedChild(const Ren
     auto availableSpace = availableAlignmentSpaceForChild(crossAxisContentExtent(), child);
-    return alignmentOffset(availableSpace, alignmentForChild(child), std::nullopt, std::nullopt, style().flexWrap() == FlexWrap::Reverse);
+    OverflowAlignment safety;
+    ItemPosition align = alignmentForChild(child, &safety);
```
Let's change this to auto along with the other local variables in the patch that are created or modified

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2356,16 +2373,18 @@ void RenderFlexibleBox::alignFlexLines(FlexLineStates& lineStates, LayoutUnit ga
 
-    if (position == ContentPosition::FlexStart && !gapBetweenLines)
+    if (position == ContentPosition::FlexStart && !gapBetweenLines
+        && !(safety == OverflowAlignment::Safe))
```
I think this could just be `safety == OverflowAlignment::Unsafe` or `safety != OverflowAlignment::Safe`

fantasai replied:
Yeah, I'm not sure why I didn't do that. :) Note that "is safe" is not the opposite of "is unsafe" because there's a "is default behavior" option as well.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2356,16 +2373,18 @@ void RenderFlexibleBox::alignFlexLines(FlexLineStates& lineStates, LayoutUnit ga
 
     ContentPosition position = style().resolvedAlignContentPosition(contentAlignmentNormalBehavior());
     ContentDistribution distribution = style().resolvedAlignContentDistribution(contentAlignmentNormalBehavior());
+    OverflowAlignment safety = style().alignContent().overflow();
```
Technically don't need a local variable here but also ok if you think it makes things clearer!

fantasai replied:
It's nice to have the parallelism... and I think maybe I had wanted to fit the next clause on one line, but forgot to unwrap it!

fantasai replied:
Is unwrapped now.

sammygill commented on Source/WebCore/rendering/RenderFlexibleBox.h:
```diff
@@ -198,7 +198,7 @@ class RenderFlexibleBox : public RenderBlock {
     void maybeCacheChildMainIntrinsicSize(RenderBox& child, bool relayoutChildren);
     void adjustAlignmentForChild(RenderBox& child, LayoutUnit);
-    ItemPosition alignmentForChild(const RenderBox& child) const;
+    ItemPosition alignmentForChild(const RenderBox& child, OverflowAlignment* safety = nullptr) const;
```
Maybe at some point we should think about just having this function return a different (new) type instead and just have the caller extract the safety value if they care about it. imo that would be a nice improvement for the api

mdubet replied:
We don't use arguments as out values generally. We should return a `std::pair` or an ad-hoc type here if necessary.

fantasai replied:
The reason I didn't do that is that [most callers](https://searchfox.org/wubkat/search?q=RenderFlexibleBox%3A%3AalignmentForChild&path=&case=true&regexp=false) don't need the safety value, and many of them are using this function inline in an expression to check the ItemPosition value--so it would be more awkward for them to use a complex return type. I can switch it over to return a synthesized StyleSelfAlignmentData or something like that, but, that's why I went with this approach.

Also, updating the return type would probably imply substantially refactoring this function, because[ it uses so many early returns](https://searchfox.org/wubkat/source/Source/WebCore/rendering/RenderFlexibleBox.cpp#2009). Idk if you want that in the same commit.

fantasai replied:
OK, went with a separate function for the oveflow alignment value. The lookup logic isn't that complicated so it should be OK to duplicate it.

mdubet commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -1813,6 +1813,11 @@ static LayoutUnit initialJustifyContentOffset(const RenderStyle& style, LayoutUn
     ContentPosition justifyContent = style.resolvedJustifyContentPosition(contentAlignmentNormalBehavior());
     ContentDistribution justifyContentDistribution = style.resolvedJustifyContentDistribution(contentAlignmentNormalBehavior());
 
+    if (availableFreeSpace < 0 && style.justifyContent().overflow() == OverflowAlignment::Safe) {
```
Why is the third condition in a separate `if` statement ?

fantasai replied:
It broke up the logic into the two logical phases: do we need to perform safe alignment, and do we need to change the value of justifyContent to make it happen.

fantasai replied:
Although, I guess technically you can't have normal safe alignment, so maybe i can just make that an assert.

mdubet commented on Source/WebCore/rendering/RenderFlexibleBox.cpp:
```diff
@@ -2313,8 +2325,13 @@ void RenderFlexibleBox::layoutColumnReverse(const FlexItems& flexItems, LayoutUn
-static LayoutUnit initialAlignContentOffset(LayoutUnit availableFreeSpace, ContentPosition alignContent, ContentDistribution alignContentDistribution, unsigned numberOfLines, bool isReversed)
+static LayoutUnit initialAlignContentOffset(LayoutUnit availableFreeSpace, ContentPosition alignContent, ContentDistribution alignContentDistribution, OverflowAlignment safety, unsigned numberOfLines, bool isReversed)
 {
+    if (availableFreeSpace < 0 && safety == OverflowAlignment::Safe) {
```
Ditto

fantasai committed:
Implement safe alignment in flexbox.
https://bugs.webkit.org/show_bug.cgi?id=263893
rdar://118000717

Reviewed by Sammy Gill.

Coverts alignment to Start when available space is negative and overflow alignment safety is on.

* LayoutTests/TestExpectations: Pass more tests!
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-002.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-003.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-004.html: Fix (and augment) incorrect test.
* LayoutTests/imported/w3c/web-platform-tests/css/css-flexbox/flexbox-safe-overflow-position-005.html: Fix (and augment) incorrect test.
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::initialJustifyContentOffset): Apply overflow safety to main-axis content alignment.
(WebCore::RenderFlexibleBox::staticCrossAxisPositionForPositionedChild): Apply overflow safety to static positioning.
(WebCore::RenderFlexibleBox::overflowAlignmentForChild const): Implement new helper method.
(WebCore::initialAlignContentOffset): Apply overflow safety to main-axis content alignment.
(WebCore::RenderFlexibleBox::alignFlexLines): Don't skip alignment calculation when safety is on, pass safety value to initialAlignContentOffset().
(WebCore::RenderFlexibleBox::alignChildren): Apply overflow safety to cross-axis item alignment.
(WebCore::RenderFlexibleBox::allowedLayoutOverflow const): Apply overflow safety to overflow bounds.
* Source/WebCore/rendering/RenderFlexibleBox.h: Add helper method overflowAlignmentForChild(), parallel to alignmentForChild().

Canonical link: https://commits.webkit.org/274304@main
